File: java/scf-bp-little-app/src/main/java/org/mydotey/scf/bp/littleapp/RawManager.java
package org.mydotey.scf.bp.littleapp;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import org.mydotey.scf.ConfigurationManager;
import org.mydotey.scf.ConfigurationManagerConfig;
import org.mydotey.scf.ConfigurationSource;
import org.mydotey.scf.Property;
import org.mydotey.scf.PropertyConfig;
import org.mydotey.scf.facade.ConfigurationManagers;
import org.mydotey.scf.facade.ConfigurationProperties;
import org.mydotey.scf.facade.StringPropertySources;
import org.mydotey.scf.source.stringproperty.propertiesfile.PropertiesFileConfigurationSourceConfig;
import org.mydotey.scf.type.string.StringToIntConverter;
import org.mydotey.scf.type.string.StringToListConverter;
import org.mydotey.scf.type.string.StringToMapConverter;

/**
 * @author koqizhao
 *
 * Jul 19, 2018
 */
public class RawManager {
    private static ConfigurationManager _manager;

    private static Property<String, String> _appId;
    private static Property<String, String> _appName;
    private static Property<String, List<String>> _userList;
    private static Property<String, Map<String, String>> _userData;
    private static Property<String, Integer> _sleepTime;
    private static Property<String, MyCustomType> _myCustomData;

    private static void initConfig() {
        // create a non-dynamic K/V (String/String) configuration source
        PropertiesFileConfigurationSourceConfig sourceConfig = StringPropertySources
                .newPropertiesFileSourceConfigBuilder().setName("app-properties-file").setFileName("app.properties")
                .build();
        ConfigurationSource source = StringPropertySources.newPropertiesFileSource(sourceConfig);

        // create a configuration manager with single source
        ConfigurationManagerConfig managerConfig = ConfigurationManagers.newConfigBuilder().setName("little-app")
                .addSource(1, source).build();
        _manager = ConfigurationManagers.newManager(managerConfig);

        // default to null
        PropertyConfig<String, String> propertyConfig = ConfigurationProperties.<String, String> newConfigBuilder()
                .setKey("app.id").setValueType(String.class).build();
        _appId = _manager.getProperty(propertyConfig);

        // default to "unknown"
        PropertyConfig<String, String> propertyConfig2 = ConfigurationProperties.<String, String> newConfigBuilder()
                .setKey("app.name").setValueType(String.class).setDefaultValue("unknown").build();
        _appName = _manager.getProperty(propertyConfig2);

        // default to empty list
        @SuppressWarnings({ "rawtypes", "unchecked" })
        PropertyConfig<String, List<String>> propertyConfig3 = ConfigurationProperties
                .<String, List<String>> newConfigBuilder().setKey("user.list").setValueType((Class) List.class)
                .setDefaultValue(new ArrayList<>()).addValueConverter(StringToListConverter.DEFAULT).build();
        _userList = _manager.getProperty(propertyConfig3);

        // default to empty map
        @SuppressWarnings({ "rawtypes", "unchecked" })
        PropertyConfig<String, Map<String, String>> propertyConfig4 = ConfigurationProperties
                .<String, Map<String, String>> newConfigBuilder().setKey("user.data").setValueType((Class) Map.class)
                .setDefaultValue(new HashMap<>()).addValueConverter(StringToMapConverter.DEFAULT).build();
        _userData = _manager.getProperty(propertyConfig4);

        // default to 1000, if value in app._properties is invalid, ignore the invalid value
        PropertyConfig<String, Integer> propertyConfig5 = ConfigurationProperties.<String, Integer> newConfigBuilder()
                .setKey("sleep.time").setValueType(Integer.class).setDefaultValue(1000)
                .addValueConverter(StringToIntConverter.DEFAULT).setValueFilter(v -> v < 0 ? null : v).build();
        _sleepTime = _manager.getProperty(propertyConfig5);

        // custom type property
        PropertyConfig<String, MyCustomType> propertyConfig6 = ConfigurationProperties
                .<String, MyCustomType> newConfigBuilder().setKey("my-custom-type-property")
                .setValueType(MyCustomType.class).addValueConverter(MyCustomType.CONVERTER).build();
        _myCustomData = _manager.getProperty(propertyConfig6);
    }

    public static void main(String[] args) throws InterruptedException {
        initConfig();

        // show properties
        System.out.println("app.id: " + _appId.getValue());
        System.out.println("app.name: " + _appName.getValue());
        System.out.println("user.list: " + _userList.getValue());
        System.out.println("user.data: " + _userData.getValue());
        System.out.println("sleep.time: " + _sleepTime.getValue());

        // get some property value for non-stable property (the key is not stable, not sure it exists or not)
        PropertyConfig<String, String> propertyConfig = ConfigurationProperties.<String, String> newConfigBuilder()
                .setKey("some.data").setValueType(String.class).setDefaultValue("not-sure").build();
        String somePropertyValue = _manager.getPropertyValue(propertyConfig);
        System.out.println();
        System.out.println("some.data: " + somePropertyValue);

        System.out.println();

        System.out.printf("Now sleep %d ms ...\n", _sleepTime.getValue());
        Thread.sleep(_sleepTime.getValue());

        System.out.println();
        System.out.println("My custom property data: " + _myCustomData.getValue());

        System.out.println();
        for (int i = 0; i < _myCustomData.getValue().getTimes(); i++)
            System.out.printf("%s %s!\n", _myCustomData.getValue().getSay(), _myCustomData.getValue().getName());

        System.out.println();
        System.out.println("Bye!");
    }

}

File: dotnet/src/MyDotey.SCF.BP.LittleApp/RawManager.cs
using System;
using System.Collections.Generic;
using System.Threading;

using MyDotey.SCF.Facade;
using MyDotey.SCF.Type.String;
using MyDotey.SCF.Source.StringProperty.PropertiesFile;

namespace MyDotey.SCF.BP.LittleApp
{
    /**
     * @author koqizhao
     *
     * Jul 19, 2018
     */
    public class RawManager
    {
        private static IConfigurationManager _manager;

        private static IProperty<string, string> _appId;
        private static IProperty<string, string> _appName;
        private static IProperty<string, List<string>> _userList;
        private static IProperty<string, Dictionary<string, string>> _userData;
        private static IProperty<string, int?> _sleepTime;
        private static IProperty<string, MyCustomType> _myCustomData;

        private static void InitConfig()
        {
            // create a non-dynamic K/V (string/string) configuration source
            PropertiesFileConfigurationSourceConfig sourceConfig = StringPropertySources
                    .NewPropertiesFileSourceConfigBuilder().SetName("app-properties-file").SetFileName("app.properties")
                    .Build();
            IConfigurationSource source = StringPropertySources.NewPropertiesFileSource(sourceConfig);

            // create a configuration manager with single source
            ConfigurationManagerConfig managerConfig = ConfigurationManagers.NewConfigBuilder().SetName("little-app")
                    .AddSource(1, source).Build();
            _manager = ConfigurationManagers.NewManager(managerConfig);

            // default to null
            PropertyConfig<string, string> propertyConfig = ConfigurationProperties.NewConfigBuilder<string, string>()
                     .SetKey("app.id").Build();
            _appId = _manager.GetProperty(propertyConfig);

            // default to "unknown"
            PropertyConfig<string, string> propertyConfig2 = ConfigurationProperties.NewConfigBuilder<string, string>()
                        .SetKey("app.name").SetDefaultValue("unknown").Build();
            _appName = _manager.GetProperty(propertyConfig2);

            // default to empty list
            PropertyConfig<string, List<string>> propertyConfig3 = ConfigurationProperties
                    .NewConfigBuilder<string, List<string>>().SetKey("user.list")
                    .SetDefaultValue(new List<string>()).AddValueConverter(StringToListConverter.Default).Build();
            _userList = _manager.GetProperty(propertyConfig3);

            // default to empty map
            PropertyConfig<string, Dictionary<string, string>> propertyConfig4 = ConfigurationProperties
                    .NewConfigBuilder<string, Dictionary<string, string>>().SetKey("user.data")
                    .SetDefaultValue(new Dictionary<string, string>()).AddValueConverter(StringToDictionaryConverter.Default).Build();
            _userData = _manager.GetProperty(propertyConfig4);

            // default to 1000, if value in app._properties is invalid, ignore the invalid value
            PropertyConfig<string, int?> propertyConfig5 = ConfigurationProperties.NewConfigBuilder<string, int?>()
                        .SetKey("sleep.time").SetDefaultValue(1000)
                        .AddValueConverter(StringToIntConverter.Default).SetValueFilter(v => v < 0 ? null : v).Build();
            _sleepTime = _manager.GetProperty(propertyConfig5);

            // custom type property
            PropertyConfig<string, MyCustomType> propertyConfig6 = ConfigurationProperties
                    .NewConfigBuilder<string, MyCustomType>().SetKey("my-custom-type-property")
                    .AddValueConverter(MyCustomType.Converter).Build();
            _myCustomData = _manager.GetProperty(propertyConfig6);
        }

        public static void DoMain(string[] args)
        {
            InitConfig();

            // show properties
            Console.WriteLine("app.id: " + _appId.Value);
            Console.WriteLine("app.name: " + _appName.Value);
            Console.WriteLine("user.list: " + _userList.Value);
            Console.WriteLine("user.data: " + _userData.Value);
            Console.WriteLine("sleep.time: " + _sleepTime.Value);

            // get some property value for non-stable property (the key is not stable, not sure it exists or not)
            PropertyConfig<string, string> propertyConfig = ConfigurationProperties.NewConfigBuilder<string, string>()
                    .SetKey("some.data").SetDefaultValue("not-sure").Build();
            string someIPropertyValue = _manager.GetPropertyValue(propertyConfig);
            Console.WriteLine();
            Console.WriteLine("some.data: " + someIPropertyValue);

            Console.WriteLine();

            Console.WriteLine("Now sleep {0} ms ...", _sleepTime.Value);
            Thread.Sleep(_sleepTime.Value.Value);

            Console.WriteLine();
            Console.WriteLine("My custom property data: " + _myCustomData.Value);

            Console.WriteLine();
            for (int i = 0; i < _myCustomData.Value.Times; i++)
                Console.WriteLine("{0} {1}!", _myCustomData.Value.Say, _myCustomData.Value.Name);

            Console.WriteLine();
            Console.WriteLine("Bye!");
        }
    }
}